### DISCUSSION TITLE
Postgres UUID type with mapped_column in 2.0b4: Could not locate core type

### DISCUSSION THREAD

**Discussion by deleted** [sqlalchemy/sqlalchemy#8980] [Usage Questions] [answered]
Hello, I'm not sure if I'm using it wrong or I've hit an edge case in the 2.0 beta, but I'm experiencing the following issue.

Using sqlalchemy `2.0.0b4` against `postgres+psycopg` (i.e. i am using psycopg 3, on purpose).  Python `3.10.5`.

I'm trying out the new declaritive syntax, and I have a model that looks like this:

```python
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import JSONB, UUID


class Base(DeclarativeBase):
    __allow_unmapped__ = True

    pass


class Namespace(Base):
    __tablename__ = "namespace"

    id: Mapped[UUID] = mapped_column(UUID, primary_key=True)
    name: Mapped[str] = mapped_column()  # TODO: citext
```

...along with some other mapped clases, all of which have UUID primary keys.

When I import the module where this is declared, I consistently get a stack trace ending in:
```
File ~/prj/chronos/.venv/lib/python3.10/site-packages/sqlalchemy/orm/properties.py:763, in MappedColumn._init_column_for_annotation(self, cls, registry, argument, originating_module)
    761         break
    762 else:
--> 763     raise sa_exc.ArgumentError(
    764         f"Could not locate SQLAlchemy Core "
    765         f"type for Python type: {our_type}"
    766     )
    768 self.column._set_type(new_sqltype)

ArgumentError: Could not locate SQLAlchemy Core type for Python type: <class 'sqlalchemy.sql.sqltypes.UUID'>
```

Am I doing something wrong? Or are postgresql dialect types not available in the v2 beta yet?

**Comment by CaselIT** [comment, marked as answer]
Hi,

The left side type needs to be a python type, so `Mapped[uuid.UUID]` instead of `Mapped[postgresql.UUID]`

@zzzeek I think we can do better for that error, by checking if the type is a `TypeEngine` indicating that it should instead be a python type

**Reply by zzzeek** [reply]
if they have the type inside of mapped_column() I'm not even sure why it would be trying to resolve it within `Mapped`

**Reply by CaselIT** [reply]
indeed it should be skipped in that case. Maybe there is also an issue here?

**Reply by zzzeek** [reply]
It's worth looking into yes, not really clear why it would be doing the lookup

**Reply by deleted** [reply]
Thanks!

Can confirm that changing the left side type to uuid.UUID prevents it from raising the ArgumentError on import.
It seems to work with or without a postgres type in the `mapped_column()` invocation.

**Comment by CaselIT** [comment]
The example as provided does not raise for me

```py
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import JSONB, UUID


class Base(DeclarativeBase):
    __allow_unmapped__ = True

    pass


class Namespace(Base):
    __tablename__ = "namespace"

    id: Mapped[UUID] = mapped_column(UUID, primary_key=True)
    name: Mapped[str] = mapped_column()  # TODO: citext
```
and the type is corretly set in the table.
When removing the type on the left side I get the  same error:
```py
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase
from sqlalchemy.dialects.postgresql import JSONB, UUID


class Base(DeclarativeBase):
    # __allow_unmapped__ = True

    pass


class Namespace(Base):
    __tablename__ = "namespace"

    id: Mapped[UUID] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column()  # TODO: citext
```
raises 
```
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type: <class 'sqlalchemy.sql.sqltypes.UUID'>
```

so there is no bug regarding what mentioned mike, but we can do better on the error side. I'll change the error message as mentioned above

**Reply by CaselIT** [reply]
New error message is at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4303
